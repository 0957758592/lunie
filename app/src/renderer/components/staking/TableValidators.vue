<template>
  <div>
    <table class="data-table">
      <thead>
        <panel-sort :sort="sort" :properties="properties" />
      </thead>
      <tbody>
        <li-validator
          v-for="validator in sortedEnrichedValidators"
          :key="validator.operator_address"
          :validator="validator"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { orderBy, isEmpty } from "lodash"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `commission`,
      order: `asc`
    },
    rollingWindow: 10000 // param of slashing period
  }),
  computed: {
    ...mapGetters([
      `delegation`,
      `committedDelegations`,
      `session`,
      `distribution`,
      `liquidAtoms`,
      `bondDenom`,
      `keybase`,
      `pool`
    ]),
    enrichedValidators() {
      return this.validators.map(v =>
        Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          percent_of_vote: v.voting_power / this.pool.pool.bonded_tokens,
          my_delegations: this.committedDelegations[v.id] > 0
            ? this.committedDelegations[v.id]
            : 0,
          commission: v.commission.rate,
          keybase: this.keybase[v.description.identity],
          rewards: this.session.signedIn
            && this.distribution.loaded && !isEmpty(this.distribution.rewards)
            ? this.distribution.rewards[v.operator_address]
            : 0,
          uptime: v.signing_info
            ? (this.rollingWindow - v.signing_info.missed_blocks_counter)
            / this.rollingWindow
            : 0
        })
      )
    },
    sortedEnrichedValidators() {
      return orderBy(
        this.enrichedValidators.slice(0),
        [this.sort.property],
        [this.sort.order]
      )
    },
    properties() {
      return [
        {
          title: `Moniker`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`
        },
        {
          title: `My Delegations`,
          value: `my_delegations`,
          tooltip: `Number of ${
            this.bondDenom
          } you have delegated to this validator`
        },
        {
          title: `Rewards`,
          value: `rewards`,
          tooltip: `Rewards you have earned from this validator`
        },
        {
          title: `Voting Power`,
          value: `percent_of_vote`,
          tooltip: `Percentage of voting shares`
        },
        {
          title: `Commission`,
          value: `commission`,
          tooltip: `The validator's commission`
        },
        {
          title: `Uptime`,
          value: `uptime`,
          tooltip: `Ratio of blocks signed within the last 10k blocks`
        },
      ]
    }
  },
  watch: {
  //   address: function() {
  //     this.session.address && this.$store.dispatch(`updateDelegates`)
  //   },
    validators: function(validators) {
      if (!validators || validators.length === 0 || !this.session.signedIn) {
        return
      }
      const yourValidators = validators.filter(
        ({ operator_address }) => operator_address in this.committedDelegations
      )

      this.$store.dispatch(`getRewardsFromAllValidators`, yourValidators)
    }
  },
  mounted() {
    this.$store.dispatch(`getPool`)
  }
}
</script>
