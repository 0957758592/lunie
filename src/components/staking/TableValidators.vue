<template>
  <div>
    <table class="data-table">
      <thead>
        <PanelSort
          :sort="sort"
          :properties="properties"
          :show-on-mobile="showOnMobile"
        />
      </thead>
      <tbody v-infinite-scroll="loadMore" infinite-scroll-distance="400">
        <LiValidator
          v-for="validator in showingValidators"
          :key="validator.operator_address"
          :validator="validator"
          :show-on-mobile="showOnMobile"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters, mapState } from "vuex"
import num from "scripts/num"
import orderBy from "lodash.orderby"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
import BN from "bignumber.js"
import { expectedReturns } from "scripts/returns"
export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `commission`,
      order: `asc`
    },
    showing: 15,
    rollingWindow: 10000 // param of slashing period
  }),
  computed: {
    ...mapState([`delegates`, `distribution`, `pool`, `session`]),
    ...mapState({
      annualProvision: state => state.minting.annualProvision
    }),
    ...mapGetters([`committedDelegations`, `bondDenom`, `lastHeader`]),
    enrichedValidators(
      {
        validators,
        delegates: { signingInfos },
        pool,
        annualProvision,
        committedDelegations,
        session,
        distribution,
        rollingWindow
      } = this
    ) {
      return validators.map(v => {
        const signingInfo = signingInfos[v.operator_address]
        return Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          my_delegations:
            session.signedIn && committedDelegations[v.operator_address] > 0
              ? committedDelegations[v.operator_address]
              : 0,
          commission: v.commission.rate,
          voting_power: BN(v.tokens)
            .div(pool.pool.bonded_tokens)
            .toFixed(10),
          rewards:
            session.signedIn && distribution.rewards[v.operator_address]
              ? distribution.rewards[v.operator_address][this.bondDenom]
              : 0,
          uptime: signingInfo
            ? (rollingWindow - signingInfo.missed_blocks_counter) /
              rollingWindow
            : 0,
          expectedReturns: annualProvision
            ? expectedReturns(
                v,
                parseInt(pool.pool.bonded_tokens),
                parseFloat(annualProvision)
              )
            : undefined
        })
      })
    },
    sortedEnrichedValidators() {
      return orderBy(
        this.enrichedValidators.slice(0),
        [this.sort.property],
        [this.sort.order]
      )
    },
    showingValidators() {
      return this.sortedEnrichedValidators.slice(0, this.showing)
    },
    properties() {
      return [
        {
          title: `Name`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`
        },
        {
          title: `Rewards`,
          value: `expectedReturns`,
          tooltip: `Approximate annualized reward`
        },
        {
          title: `Voting Power`,
          value: `voting_power`,
          tooltip: `Percentage of voting shares`
        }
      ]
    }
  },
  watch: {
    lastHeader: {
      immediate: true,
      handler() {
        this.$store.dispatch(`getRewardsFromMyValidators`)
      }
    },
    "sort.property": function() {
      this.showing = 15
    },
    "sort.order": function() {
      this.showing = 15
    }
  },
  mounted() {
    this.$store.dispatch(`getPool`)
    this.$store.dispatch(`getRewardsFromMyValidators`)
    this.$store.dispatch(`getMintingParameters`)
  },
  methods: {
    loadMore() {
      this.showing += 10
    }
  }
}
</script>
<style scoped>
@media screen and (max-width: 550px) {
  .data-table td {
    overflow: hidden;
  }

  .data-table__row__info {
    max-width: 22rem;
  }
}
</style>
