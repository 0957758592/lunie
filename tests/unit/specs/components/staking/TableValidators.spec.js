import { shallowMount } from "@vue/test-utils"
import TableValidators from "src/components/staking/TableValidators"
// import validators from "../../store/json/validators.js"

const validators = [
  {moniker: 'cosmos1a', operatorAddress: 'cosmos1a'},
  {moniker: 'cosmos1b', operatorAddress: 'cosmos1b'},
  {moniker: 'cosmos1c', operatorAddress: 'cosmos1c'},
  {moniker: 'cosmos1d', operatorAddress: 'cosmos1d'},
  {moniker: 'cosmos1e', operatorAddress: 'cosmos1e'},
  {moniker: 'cosmos1f', operatorAddress: 'cosmos1f'},
  {moniker: 'cosmos1g', operatorAddress: 'cosmos1g'},
  {moniker: 'cosmos1h', operatorAddress: 'cosmos1h'},
  {moniker: 'cosmos1i', operatorAddress: 'cosmos1i'},
  {moniker: 'cosmos1j', operatorAddress: 'cosmos1j'},
  {moniker: 'cosmos1k', operatorAddress: 'cosmos1k'},
  {moniker: 'cosmos1l', operatorAddress: 'cosmos1l'},
  {moniker: 'cosmos1m', operatorAddress: 'cosmos1m'},
  {moniker: 'cosmos1n', operatorAddress: 'cosmos1n'},
  {moniker: 'cosmos1o', operatorAddress: 'cosmos1o'},
]

describe(`TableValidators`, () => {
  let wrapper


  beforeEach(() => {
    wrapper = shallowMount(TableValidators, {
      propsData: { validators },
      directives: {
        infiniteScroll: () => {}
      }
    })
  })

  it(`shows a validator table`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should add small_moniker property to validators`, () => {
    expect(wrapper.vm.showingValidators[0].small_moniker).toBe(`cosmos1a`)
  })

  it(`should sort the delegates by selected property`, () => {
    wrapper.vm.sort.property = `operatorAddress`
    wrapper.vm.sort.order = `desc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operatorAddress)
    ).toEqual(validators.map(x => x.operatorAddress).reverse())

    wrapper.vm.sort.property = `operatorAddress`
    wrapper.vm.sort.order = `asc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operatorAddress)
    ).toEqual(validators.map(x => x.operatorAddress))
  })

  it(`should load more validators (infinite scrolling)`, async () => {
    wrapper.setData({ showing: 2 })
    expect(wrapper.findAll("livalidator-stub").length).toBe(2)
    wrapper.vm.loadMore()
    expect(wrapper.findAll("livalidator-stub").length).toBe(12)
  })

})
