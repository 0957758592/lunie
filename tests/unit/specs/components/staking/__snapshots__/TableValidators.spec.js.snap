// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableValidators shows a validator table 1`] = `
<div>
  <table
    class="data-table"
  >
    <thead>
      <panelsort-stub
        properties="[object Object],[object Object],[object Object]"
        showonmobile="returns"
        sort="[object Object]"
      />
    </thead>
     
    <tbody
      infinite-scroll-distance="400"
      name="flip-list"
    >
      <livalidator-stub
        index="0"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="1"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="2"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="3"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="4"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="5"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="6"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="7"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="8"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="9"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="10"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="11"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="12"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="13"
        showonmobile="returns"
        validator="[object Object]"
      />
      <livalidator-stub
        index="14"
        showonmobile="returns"
        validator="[object Object]"
      />
    </tbody>
  </table>
</div>
`;
